File: src/sort/MergeSort.java
package sort;

/* 
 * Sample program for Merge Sort algorithm.
 * 
 * https://geeksforgeeks.org/merge-sort/ 
 * https://visualgo.net/en/sorting
 * 
 * There are two phases: splitting and merging. The Splitting phase leads to faster sorting during the Merging phase. Splitting
 * is logical. We don't create new arrays. The array is split into two until the arrays have only one element - these are sorted.
 * Then, merge every left/right pair of sibling arrays into a sorted array
 *
 * O(nLog(n))
 * Divide and conquer
 * Recursive
 * Not in place
 * Stable algorithm
 */

public class MergeSort {
	// Merges two subarrays of arr[].
	// First subarray is arr[l..m]
	// Second subarray is arr[m+1..r]
	void merge(int arr[], int l, int m, int r)
	{
		// Find sizes of two subarrays to be merged
		int n1 = m - l + 1;
		int n2 = r - m;

		/* Create temp arrays */
		int L[] = new int [n1];
		int R[] = new int [n2];

		/*Copy data to temp arrays*/
		for (int i=0; i<n1; ++i)
			L[i] = arr[l + i];
		for (int j=0; j<n2; ++j)
			R[j] = arr[m + 1+ j];

		/* Merge the temp arrays */

		// Initial indexes of first and second subarrays
		int i = 0, j = 0;

		// Initial index of merged subarray
		int k = l;
		while (i < n1 && j < n2)
		{
			if (L[i] <= R[j])
			{
				arr[k] = L[i];
				i++;
			}
			else
			{
				arr[k] = R[j];
				j++;
			}
			k++;
		}

		/* Copy remaining elements of L[] if any */
		while (i < n1)
		{
			arr[k] = L[i];
			i++;
			k++;
		}

		/* Copy remaining elements of R[] if any */
		while (j < n2)
		{
			arr[k] = R[j];
			j++;
			k++;
		}
	}

	// Main function that sorts arr[l..r] using
	// merge()
	void sort(int arr[], int l, int r)
	{
		if (l < r)
		{
			// Find the middle point
			int m = (l+r)/2;

			// Sort first and second halves
			sort(arr, l, m);
			sort(arr , m+1, r);

			// Merge the sorted halves
			merge(arr, l, m, r);
		}
	}

	/* A utility function to print array of size n */
	static void printArray(int arr[])
	{
		int n = arr.length;
		for (int i=0; i<n; ++i)
			System.out.print(arr[i] + " ");
		System.out.println();
	}

	// Driver method
	public static void main(String args[])
	{
		int arr[] = {12, 11, 13, 5, 6, 7};

		System.out.println("Given Array");
		printArray(arr);

		MergeSort ob = new MergeSort();
		ob.sort(arr, 0, arr.length-1);

		System.out.println("\nSorted array");
		printArray(arr);
	}

}

File: src/c++/sort/MergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> merge_sort(vector<int> arr);

/*
 * MERGE SORT ALGORITHM
 *
 * This sort algorithm using Divide and Conquer. The array is recursively split in half until all subarrays have 1 or 0 elements.
 * Then each of these subarrays is merged using the merging algorithm.
 *
 * O(NlogN), stable, comparison, NOT inplace, recursion, merging, divide and conquer
 * 
 */

vector<int> merge_sort(vector<int> arr) {
    // BASE CASE
    if (arr.size() <= 1) return arr;

    // DIVIDE STEP
    int mid = arr.size() / 2;
    vector<int> arr1, arr2;
    for (int i = 0; i < mid; i++) arr1.push_back(arr[i]);
    for (int i = mid; i < arr.size(); i++) arr2.push_back(arr[i]);

    // RECURSE STEP
    vector<int> sorted_arr1 = merge_sort(arr1);
    vector<int> sorted_arr2 = merge_sort(arr2);

    // MERGE/RECOMBINE STEP
    vector<int> sorted_arr = {};
    int i1 = 0;
    int i2 = 0;
    while (i1 < sorted_arr1.size() && i2 < sorted_arr2.size()) {
        if (sorted_arr1[i1] <= sorted_arr2[i2]) {
            sorted_arr.push_back(sorted_arr1[i1]);
            i1++;
        } else {
            sorted_arr.push_back(sorted_arr2[i2]);
            i2++;
        }
    }
    while (i1 < sorted_arr1.size()) {
        sorted_arr.push_back(sorted_arr1[i1]);
        i1++;
    }
    while (i2 < sorted_arr2.size()) {
        sorted_arr.push_back(sorted_arr2[i2]);
        i2++;
    }

    return sorted_arr;
}

int main() {
    vector<int> arr = {6, 3, -1, 0, 4, 2, 1, 3, 5}; // -1 0 1 2 3 3 4 5 6
    vector<int> sorted = merge_sort(arr);
    for (int i : sorted) {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}